Build history pages from content item

KludgeKML opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Allows pages under /government/history to be built from the new history documents that will come out of Whitehall (based on their standard edition).

- adds support for /government/history/:slug to render the location history pages currently on government-frontend (eg: https://www.gov.uk/government/history/10-downing-street).
- adds support for rendering /government/history via a little trick (we respond to that route, but intercept it at content item loader time and load the content item from /government/history/history-of-the-uk-government), since Whitehall doesn't support arbitrary paths or heirarchical paths at the moment. We can remove this hack when Whitehall supports those paths fully. 

## Why

https://trello.com/c/pMEJlBME/796-build-history-pages-from-content-item, [Jira issue PNP-5622](https://gov-uk.atlassian.net/browse/PNP-5622)

## Screenshots

Location history pages change very little (just a change in the spacing at the top), but the main history homepage changes a lot (from the zig-zag before to a content list/content column pair). This has been agreed with our designer.

Before|After
------|------
<img width="2455" height="3404" alt="image" src="https://github.com/user-attachments/assets/0bf106da-714c-4227-b73e-ac281aee6290" />|<img width="2455" height="5159" alt="image" src="https://github.com/user-attachments/assets/0ab9c976-b7aa-484f-8e46-4dab9587ac4a" />
<img width="2455" height="19841" alt="image" src="https://github.com/user-attachments/assets/d6679263-6ec3-4529-938b-09eecc660052" />|<img width="2455" height="19806" alt="image" src="https://github.com/user-attachments/assets/ad8c281e-8fbb-488a-9e7d-86c3d106eee7" />

KludgeKML committed:
feat: Allow ContentItemFactory to build sub-sub-classes of ContentItem

- We want to allow models that are subclasses of FlexiblePage, that can supply default
  flexible_section configurations in the absence of a specified one. Before the factory
  would default to a ContentItem if the detected class wasn't ContentItem or a direct
  subclass of ContentItem.
- Uses the <= operator, which on a class checks if the class is a subclass of the
  comparator (or a subclass of that, etc). https://ruby-doc.org/3.4.1/Module.html#method-i-3C-3D

KludgeKML committed:
feat: Allow flexible page constructor to defer default sections

- For the new flexible page types some will come with flexible section information in the
  content item, but some will not (like the new type of history page). For those we want
  to be able to supply a default flexible section layout. The simplest initial idea is for any
  subclass of FlexiblePage to supply its own configuration array, filled in with the appropriate
  details
- The intent is that relatively fixed types of flexible page will override default_flexible_sections
  (an example will be in the following commit) to provide a flexible_sections array made up from
  data in the content_item. But if the content item comes with its own flexible_sections array,
  that will be used instead. This allows us to potentially make one-off flexible pages that otherwise
  look like a history page, but have a wildly different layout. This will need Whitehall support,
  though, so for the moment is just a possibility.
- We also deep_stringify_keys on the hash before we pass it to the factory so that developers can
  use modern hash syntax in the models rather than string keys (which would otherwise be necessary
  to match the keys coming out of the GdsApi calls that build the content item)

KludgeKML committed:
fix: remove layout-level breadcrumbs for history

- Like all flexible pages, history will specify its own breadcrumbs, so we explicitly
  opt it out from the contextual breadcrumbs usually added in the layout.
- Breadcrumbs in frontend should probably be refactored, it's not clear that the
  contextual breadcrumbs component should really exist (it would be better just to
  explicitly specify breadcrumbs in the rendering app and not have a component that
  has to know about content item structure). But that's out of the scope of this
  commit/PR.

KludgeKML committed:
fix: add govuk-!-width-full to rich content sidebar image

- We may not always get perfect sized images for this sidebar header image, but we want it to be
  the same size as the column its in.

KludgeKML committed:
feat: add HubRedirectLoader

- In order to move the /government/history hub page off government-frontend, we need to
  support that path. But the current publishing system for history pages on Whitehall doesn't
  support arbitrary or heirarchical paths - all history items are off /government/history. As a
  temporary workaround to get history pages out of government-frontend, we've decided to
  allow frontend to treat /government/history as a special case. We serve it as is, but we
  retrieve the content item from /government/history/history-of-the-uk-government.

- This ContentItemLoader adds support for that. It returns true from #can_load? if the path
  matches /government/history, and when asked to load that path it loads the specified
  content item from the content store instead and returns that.

- We add it to the alternative loaders list after the file-based loaders. Its behaviour isn't
  guaranteed if it goes after the Graphql loader.

KludgeKML committed:
fix: remove machine-readable metadata for flexible pages

- history page doesn't need this, and if it's included it increases the page size by a lot
  by including the whole body of the article in the article JSON. We may reinstate this later
  for other flexible pages, but for now remove it.
- Note that this is the JSON-LD machine-readable metadata as per
  https://components.publishing.service.gov.uk/component-guide/machine_readable_metadata,
  not the metadata tags in the header.

KludgeKML committed:
test: add test for no image in rich content flexible section

KludgeKML committed:
feat: specify breadcrumbs for flexible pages (and history pages)

- We assume all flexible pages will at least have a home breadcrumb,
  and override that in the history model to add /government/history as the second
  breadcrumb, matching the current files.

AshGDS commented on app/models/history.rb:
```diff
@@ -0,0 +1,29 @@
+      alt: content_store_response.dig("details", "sidebar_image", "caption"),
+      src: content_store_response.dig("details", "sidebar_image", "url"),
+    }
+  end
```
Is it more efficient to store  the `content_store_response.dig("details", "sidebar_image")` part in a variable here so that it's only called once?

AshGDS reviewed (commented):
Was just wondering where/when it was agreed for the History of the UK Government to have its layout changed? I'm guessing it's a compromise to make the pages editable now?

And not sure if I'm doing something wrong, but the history pages do not load properly for me locally - `/government/history` 404s, and the No 10 page returns this:

<img width="711" height="534" alt="image" src="https://github.com/user-attachments/assets/08235ee7-1363-4fb7-acfe-85964f219ed4" />

(Code looks good though 👍 )

KludgeKML commented:
> Was just wondering where/when it was agreed for the History of the UK Government to have its layout changed? I'm guessing it's a compromise to make the pages editable now?

Yeah, we discussed it with Nikin. The zig-zag styling wasn't awful, but it isn't something we've done anywhere else, and this lets us standardise and clear all the old hardcoded history pages off government-frontend for low effort.

> And not sure if I'm doing something wrong, but the history pages do not load properly for me locally - `/government/history` 404s, and the No 10 page returns this:

Ah, you need to be pointing your local machine at integration, not prod.

KludgeKML committed:
feat: Add flexible page rendering for /government/history files

- Adds route for /government/history/:slug pointing to the flexible page controller
- Adds a History model, derived from FlexiblePage, which will supply a default
  flexible-pages section config of a page_title derived from the content item's title
  and optionally lead_paragraph, and a rich_content derived from the content items's
  body, headings, and optionally sidebar_image.
- This corresponds to the output from the history_page configurable document type
  https://github.com/alphagov/whitehall/blob/3ca07751d7fe1cab5aadd9ddbb777ed1dde00e36/app/models/configurable_document_types/history_page.json
- ..and the history content schema
  https://github.com/alphagov/publishing-api/blob/97b26a381048ed3a526f97f75f1e1206b210955e/content_schemas/formats/history.jsonnet

AshGDS reviewed (approved):
LGTM 👍  Two things to note:

- On the Number 10 page, the `first-published-at` etc. dates are now 2025, do we need to preserve the original dates?
- The GA4 indexes shift slightly on the Number 10 page, so we probably need to let the performance analysts know when we publish this page.

deborahchua reviewed (approved):
LGTM!
